### DISCUSSION TITLE
What causes "UNABLE_TO_VERIFY_LEAF_SIGNATURE"?

### DISCUSSION THREAD

**Discussion by crystalfp** [FiloSottile/mkcert#490] [Q&A]
First time using mkcert for my local server:
```
./mkcert-v1.4.4-windows-amd64.exe \
-cert-file "./localhost-cert.pem" \
-key-file "./localhost-privkey.pem" \
localhost 127.0.0.1 ::1
```
Accessing the server from my frontend, that uses `fetch()`, works perfectly. Instead, accessing the server from [Pactum ](https://pactumjs.github.io/) (API tester) generates the "UNABLE_TO_VERIFY_LEAF_SIGNATURE" error.
Also changing the access routine Pactum uses with the following:
```
const https = require('https');
const pem = 'path-pem-file';
const key = 'path-key-file';

const agent = new https.Agent({
            cert: fs.readFileSync(pem),
            key: fs.readFileSync(key), 
	    passphrase: ""
 });
```
changes nothing. This led me to suspect I'm missing something during mkcert invocation.
Could you enlighten me?
In the meantime I will investigate how Pactum access the server.
Thanks!
mario

**Comment by crystalfp** [comment]
Also adding `-client` to mkcert changes nothing.

**Reply by timostamm** [reply]
Mario, the relevant info is in the [README](https://github.com/FiloSottile/mkcert#using-the-root-with-nodejs):

> Node does not use the system root store, so it won't accept mkcert certificates automatically. Instead, you will have to set the [NODE_EXTRA_CA_CERTS](https://nodejs.org/api/cli.html#cli_node_extra_ca_certs_file) environment variable.
> ```
> export NODE_EXTRA_CA_CERTS="$(mkcert -CAROOT)/rootCA.pem"
> ```

**Reply by crystalfp** [reply]
Thanks @timostamm . But unfortunately nothing changes, even if I do `process.env.NODE_EXTRA_CA_CERTS = '["C:/Users/xxx/AppData/Local/mkcert/RootCA.pem"]';` using the directory from `mkcert -CAROOT`.

Summarizing: https node server ← browser works. Instead https node server ← jest test using pactum fails with UNABLE_TO_VERIFY_LEAF_SIGNATURE. Anyway, seems I need to learn more about HTTP/2. In the meantime my tests will continue to run on http.
Thanks anyway!
mario

**Reply by timostamm** [reply]
From the Node.js docs:
> The NODE_EXTRA_CA_CERTS environment variable is only read when the Node.js process is first launched. Changing the value at runtime using process.env.NODE_EXTRA_CA_CERTS has no effect on the current process.

You have to set this environment variable _before_ starting node. You can't do it in the process.  :upside_down_face:

Here's a full example: https://github.com/bufbuild/connect-es/tree/main/packages/example#run-the-example

**Reply by crystalfp** [reply]
Yes, I defined also the Env variable. And thanks for the example, I will study it.
